freecad: add freecad-wayland to fix native wayland

ppenguin opened the pull request:
## Description of changes

The current `FreeCAD` in `nixpkgs` would only run on `xwayland` when running in a `wayland` environment.
This PR adds a `freecad-wayland` attribute that runs natively on `wayland` (i.e. no `xwayland`). It is for now done under a separate attribute to minimise the impact on the existing package (since the wrapper sets the `QPA_PLATFORM` on the original package).

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented:
@ofborg build freecad-wayland

AndersonTorres commented:
Nothing has passed on ofborg.

ppenguin commented:
> Nothing has passed on ofborg.

Was it hanging?
It looks like now all has passed (or skipped)

ppenguin commented:
Ah, now it seems green, so somebody with write access could merge this?
Cheers!

AndersonTorres commented on pkgs/applications/graphics/freecad/default.nix:
```diff
@@ -140,7 +143,7 @@ stdenv.mkDerivation (finalAttrs: {
     "--set COIN_GL_NO_CURRENT_CONTEXT_CHECK 1"
     "--prefix PATH : ${libredwg}/bin"
-    "--set QT_QPA_PLATFORM xcb"
+    (lib.optional (!withWayland) "--set QT_QPA_PLATFORM xcb")
   ];
```
Nitty:

```suggestion
  ] ++ lib.optionals (!withWayland) [ "--set QT_QPA_PLATFORM xcb" ];
```

AndersonTorres commented:
Squash the commits.

ppenguin commented:
> Squash the commits.

done

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -31967,6 +31967,8 @@ with pkgs;
       shiboken2;
   };
 
+  freecad-wayland = pkgs.freecad.override { withWayland = true; };
```
```suggestion
  freecad-wayland = freecad.override { withWayland = true; };
```

ppenguin replied:
ah yes, of course...

ppenguin committed:
freecad: add freecad-wayland that runs on wayland natively

freecad: add freecad-wayland that runs on wayland natively (no xwayland)

reformat

all-packages

fix freecad

MaxwellDupre reviewed (approved):
LGTM

ppenguin commented:
> Fix the conflicts

Done (finally)

AndersonTorres commented:
Why the so huge diff in all-packages.nix?

AndersonTorres commented on pkgs/by-name/fr/freecad/package.nix:
```diff
@@ -54,144 +58,151 @@ let
+        right at home with FreeCAD.
+      '';
+      license = lib.licenses.lgpl2Plus;
+      maintainers = with lib.maintainers; [viric gebner AndersonTorres];
```
What formatter are you using?

ppenguin replied:
Oops, sorry I missed that one, that was unintended! It appears to be the default formatter that `nvf` is using... I'll reconfig and revert that part...

SuperSandro2000 reviewed (changes requested):
Please undo the formatting changes in this PR. We want to keep the previous formatting done with the nixfmt rfc most likely.

ppenguin commented:
> Please undo the formatting changes in this PR. We want to keep the previous formatting done with the nixfmt rfc most likely.

Done (was unintended, caused by a wrong nvf/nvim default setting).
